### PR TITLE
=art #20455 compression tables advertised as a whole "2 tables mode"

### PR DESCRIPTION
**Pull request by ktoso**
Implements "2 table mode" for Artery Compression, also known as "we use one table and advertise the next one, on first receive of the new table, we drop the old one and start using the fresh one".

Will want to clean up a bit and add more tests to cover corner cases, but this is a working version already. Includes test for "old table be dropped for new system" etc.
- [ ] check log levels.
- [ ] re-visit TODOs
- [ ] possible array pooling in InboundCompression 
- [ ] make sure we don't call resolveActorRef more than needed

Wording has been changed somewhat "Table" is the actual thing we send over the wire, and "Compressions" is the active entity in which we compress actorRefs / manifests etc. It contains a Table.

### CODE REVIEW

File: akka-remote/src/main/scala/akka/remote/artery/Codecs.scala [hunk 1]
```diff
@@ -176,14 +182,14 @@ private[remote] class Decoder(
             // recipient for the first message that is sent to it, best effort retry
             scheduleOnce(RetryResolveRemoteDeployedRecipient(
               retryResolveRemoteDeployedRecipientAttempts,
-              headerBuilder.recipientActorRefPath, decoded), retryResolveRemoteDeployedRecipientInterval)
+              headerBuilder.recipientActorRefPath.get, decoded), retryResolveRemoteDeployedRecipientInterval) // FIXME IS THIS SAFE?
```
Comment (patriknw) [inline akka-remote/src/main/scala/akka/remote/artery/Codecs.scala]: yes, that should be safe, because of `!headerBuilder.isNoRecipient`

File: akka-remote/src/test/scala/akka/remote/artery/compress/CompressionTableSpec.scala [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+ */
+
+package akka.remote.artery.compress
+
+import akka.testkit.AkkaSpec
+
+class CompressionTableSpec extends AkkaSpec {
+
+  "CompressionTable" must {
+    "should invert" in {
```
Comment (patriknw) [inline akka-remote/src/test/scala/akka/remote/artery/compress/CompressionTableSpec.scala]: must should

File: akka-remote-tests/src/multi-jvm/scala/akka/remote/artery/MaxThroughputSpec.scala [hunk 3]
```diff
@@ -55,6 +54,12 @@ object MaxThroughputSpec extends MultiNodeConfig {
...
            #advanced.embedded-media-driver = off
            #advanced.aeron-dir = "target/aeron"
+           
+           #advanced.compression {
+           #  enabled = on
+           #  actor-refs.enabled = on
+           #  manifests.enabled = on
+           #}
          }
        }
```
Comment (patriknw) [inline akka-remote-tests/src/multi-jvm/scala/akka/remote/artery/MaxThroughputSpec.scala]: how does the advertisement work for this test? we want to run benchmarks when tables are in place

**Comment by patriknw** [comment]
LGTM

File: akka-remote/src/test/scala/akka/remote/artery/compress/CompressionTableSpec.scala (hunk 2 above)
Comment (ktoso) [inline akka-remote/src/test/scala/akka/remote/artery/compress/CompressionTableSpec.scala]: thanks

File: akka-remote-tests/src/multi-jvm/scala/akka/remote/artery/MaxThroughputSpec.scala (hunk 3 above)
Comment (ktoso) [inline akka-remote-tests/src/multi-jvm/scala/akka/remote/artery/MaxThroughputSpec.scala]: Advertised table is:

```
[InboundCompressionTable(akka://MaxThroughputSpec)] Advertise ActorRef compression [CompressionTable(1,
Map(Actor[artery://MaxThroughputSpec@localhost:58601/user/5-to-5-snd4#-798311822] -> 9, 
Actor[akka://MaxThroughputSpec/user/5-to-5-rcv4#133185920] -> 8, 
Actor[artery://MaxThroughputSpec@localhost:58601/system/testActor-1#7166865] -> 1, 
Actor[akka://MaxThroughputSpec/user/5-to-5-rcv1#-88748461] -> 2, 
Actor[akka://MaxThroughputSpec/user/5-to-5-rcv5#1348321425] -> 10, 
Actor[akka://MaxThroughputSpec/user/5-to-5-rcv2#-1319771620] -> 4, 
Actor[artery://MaxThroughputSpec@localhost:58601/user/5-to-5-snd3#-1135029393] -> 7, 
Actor[akka://MaxThroughputSpec/user/5-to-5-rcv3#865004274] -> 6, 
Actor[artery://MaxThroughputSpec@localhost:58601/user/5-to-5-snd5#499982012] -> 11, 
Actor[artery://MaxThroughputSpec@localhost:58601/user/5-to-5-snd2#647171527] -> 5, 
Actor[artery://MaxThroughputSpec@localhost:58601/user/5-to-5-snd1#1751833096] -> 3, 
Actor[akka://MaxThroughputSpec/deadLetters] -> 0
))] to [artery://MaxThroughputSpec@localhost:58601]
```

It does slow down currently, so I'll work on that next (addressing many of the TODOs etc):

with: `[JVM-1] === MaxThroughput 5-to-5-snd5: throughput 96,177 msg/s, 9,617,697 bytes/s, dropped 0, max round-trip 116 ms, burst size 200, payload size 100, 20795 ms to deliver 400000 messages`

without: `[JVM-1] === MaxThroughput 5-to-5-snd5: throughput 171,821 msg/s, 17,182,131 bytes/s, dropped 0, max round-trip 109 ms, burst size 200, payload size 100, 11640 ms to deliver 400000 messages`

I'll work on profiling now

Comment (patriknw) [inline akka-remote-tests/src/multi-jvm/scala/akka/remote/artery/MaxThroughputSpec.scala]: and it was a major drop in the first table PR, we should get back to where we were. Potentially it should be better (we send less data) than before. See build 145:
https://jenkins.akka.io:8498/job/akka-artery-perf/plot/MaxThroughputSpec/
https://jenkins.akka.io:8498/job/akka-artery-perf/plot/LatencySpec/

Comment (patriknw) [inline akka-remote-tests/src/multi-jvm/scala/akka/remote/artery/MaxThroughputSpec.scala]: my question was more about how do we warmup and only start the real measurements when tables have been advertised?

Comment (ktoso) [inline akka-remote-tests/src/multi-jvm/scala/akka/remote/artery/MaxThroughputSpec.scala]: Ah, no that we don't have yet.
The perf flamegraphs actually do the right thing here, since they attach after 5 seconds, but the notmal measurements in the test don't (yet)

**Comment by ktoso** [comment]
Resolved https://github.com/akka/akka/issues/20729

**Comment by ktoso** [comment]
Resolved https://github.com/akka/akka/issues/20734
